File: csharp/robot-simulator/RobotSimulator.cs
public enum Direction
{
    North,
    East,
    South,
    West
}

public class RobotSimulator(Direction direction, int x, int y)
{
    public Direction Direction { get; private set; } = direction;

    public int X { get; private set; } = x;

    public int Y { get; private set; } = y;

    public void Move(string instructions)
    {
        foreach (var instruction in instructions)
        {
            switch (instruction)
            {
                case 'A':
                    switch (Direction)
                    {
                        case Direction.North: Y++; break;
                        case Direction.East: X++; break;
                        case Direction.South: Y--; break;
                        case Direction.West: X--; break;
                    }
                    break;
                case 'L':
                    Direction = (Direction)(((int)Direction - 1 + 4) % 4);
                    break;
                case 'R':
                    Direction = (Direction)(((int)Direction + 1) % 4);
                    break;
            }
        }
    }
}

File: cpp/robot-simulator/robot_simulator.cpp
#include "robot_simulator.h"

namespace robot_simulator
{

    Robot::Robot(std::pair<int, int> position, Bearing bearing) : position(position), bearing(bearing) {}

    std::pair<int, int> Robot::get_position() const
    {
        return position;
    }

    Bearing Robot::get_bearing() const
    {
        return bearing;
    }

    void Robot::turn_left()
    {
        bearing = (Bearing)(((int)bearing - 1 + 4) % 4);
    }

    void Robot::turn_right()
    {
        bearing = (Bearing)(((int)bearing + 1) % 4);
    }

    void Robot::advance()
    {
        int x = position.first;
        int y = position.second;
        switch (bearing)
        {
        case Bearing::NORTH:
            y++;
            break;
        case Bearing::EAST:
            x++;
            break;
        case Bearing::SOUTH:
            y--;
            break;
        case Bearing::WEST:
            x--;
            break;
        }

        position = {x, y};
    }

    void Robot::execute_sequence(const std::string &instructions)
    {
        for (auto instruction : instructions)
        {
            switch (instruction)
            {
            case 'A':
                advance();
                break;
            case 'L':
                turn_left();
                break;
            case 'R':
                turn_right();
                break;
            }
        }
    }
} // namespace robot_simulator
